### PR TITLE
dcos-1456 add 'New command(s) available' message for install

### PR DESCRIPTION
**Pull request by tamarrow**

### CODE REVIEW

File: dcos/subcommand.py [hunk 1]
```diff
@@ -14,7 +14,7 @@
 def command_executables(subcommand, dcos_path):
     """List the real path to executable dcos program for specified subcommand.
 
-    :param subcommand: name of subcommand. E.g. marathon
+    :param subcommand: name of subcommand. E.g. Marathon
```
Comment (jsancio) [inline dcos/subcommand.py]: The example should be in lower case, no? The package name is `marathon` not `Marathon` I think.

Comment (tamarrow) [inline dcos/subcommand.py]: true, will change back

File: dcos/subcommand.py [hunk 2]
```diff
@@ -39,13 +39,39 @@ def command_executables(subcommand, dcos_path):
     return executables[0]
 
 
+def list_package_exectuables(package, dcos_path):
```
Comment (mgummelt) [inline dcos/subcommand.py]: s/package/package_name

File: dcos/subcommand.py [hunk 3]
```diff
@@ -39,13 +39,39 @@ def command_executables(subcommand, dcos_path):
...
+    """List the real path(s) to executables for a specific dcos subcommand
```
Comment (mgummelt) [inline dcos/subcommand.py]: what is a "real path"?

Comment (tamarrow) [inline dcos/subcommand.py]: absolute path

File: dcos/subcommand.py [hunk 4]
```diff
@@ -39,13 +39,39 @@ def command_executables(subcommand, dcos_path):
...
+    :param dcos_path: path to the dcos cli directory
+    :type dcos_path: str
+    :returns: list of all the dcos program paths in package
+    :rtype: [str]
+    """
+    bin_dir = os.path.join(package_dir(package),
+                           constants.DCOS_SUBCOMMAND_VIRTUALENV_SUBDIR,
+                           BIN_DIRECTORY)
+
+    subcommands = []
```
Comment (mgummelt) [inline dcos/subcommand.py]: s/subcommands/executables

File: dcos/subcommand.py (hunk 2 above)
Comment (mgummelt) [inline dcos/subcommand.py]: s/list/get

File: dcos/subcommand.py [hunk 5]
```diff
@@ -39,13 +39,39 @@ def command_executables(subcommand, dcos_path):
...
     """List the real path to executable dcos subcommand programs.
 
     :param dcos_path: path to the dcos cli directory
     :type dcos_path: str
     :returns: list of all the dcos program paths
-    :rtype: list of str
+    :rtype: [str]
     """
 
     # Let's get all the default subcommands
```
Comment (mgummelt) [inline dcos/subcommand.py]: looks like some code is repeated here and in the new function above

File: dcos/subcommand.py (hunk 2 above)
Comment (mgummelt) [inline dcos/subcommand.py]: this function doesn't get all executables.  it only gets those that are dcos commands.  So maybe call this "get_package_commands"

File: cli/dcoscli/package/main.py [hunk 6]
```diff
@@ -348,6 +348,19 @@ def _install(package_name, options_path, app_id, cli, app, yes):
 
         subcommand.install(pkg, pkg_version, options)
 
+        name = pkg.name()
+        path = util.dcos_path()
+        subcommand_paths = subcommand.list_package_exectuables(name, path)
+        new_commands = []
+        for path in subcommand_paths:
+            subcommand_name = path.split('/')[-1]
```
Comment (mgummelt) [inline cli/dcoscli/package/main.py]: use `os.path.basename`

File: cli/dcoscli/package/main.py [hunk 7]
```diff
@@ -348,6 +348,19 @@ def _install(package_name, options_path, app_id, cli, app, yes):
 
         subcommand.install(pkg, pkg_version, options)
 
+        name = pkg.name()
+        path = util.dcos_path()
+        subcommand_paths = subcommand.list_package_exectuables(name, path)
+        new_commands = []
```
Comment (mgummelt) [inline cli/dcoscli/package/main.py]: you should be able to do this in a list comprehension

File: cli/dcoscli/package/main.py [hunk 8]
```diff
@@ -348,6 +348,19 @@ def _install(package_name, options_path, app_id, cli, app, yes):
...
+        for path in subcommand_paths:
+            subcommand_name = path.split('/')[-1]
+            subcommand_name = subcommand_name.replace('-', ' ', 1)
+            new_commands.append(subcommand_name)
+
+        if len(new_commands) > 0:
+            commands = ', '.join(new_commands)
+            emitter.publish("New command(s) available: {}".format(commands))
```
Comment (mgummelt) [inline cli/dcoscli/package/main.py]: Let's present them as the user would type them (e.g. "dcos kafka", not "kafka")

File: cli/dcoscli/package/main.py [hunk 9]
```diff
@@ -348,6 +348,19 @@ def _install(package_name, options_path, app_id, cli, app, yes):
 
         subcommand.install(pkg, pkg_version, options)
 
+        name = pkg.name()
```
Comment (mgummelt) [inline cli/dcoscli/package/main.py]: you already have a `package_name` variable

File: dcos/subcommand.py (hunk 2 above)
Comment (mgummelt) [inline dcos/subcommand.py]: let's not pass around dcos_path.  we can get it cheaply via util.dcos_path()

File: cli/dcoscli/package/main.py [hunk 10]
```diff
@@ -348,6 +348,19 @@ def _install(package_name, options_path, app_id, cli, app, yes):
...
+        path = util.dcos_path()
+        subcommand_paths = subcommand.list_package_exectuables(name, path)
+        new_commands = []
+        for path in subcommand_paths:
+            subcommand_name = path.split('/')[-1]
+            subcommand_name = subcommand_name.replace('-', ' ', 1)
+            new_commands.append(subcommand_name)
+
+        if len(new_commands) > 0:
```
Comment (mgummelt) [inline cli/dcoscli/package/main.py]: `if new_commands:` is the "pythonic" way of doing this

File: cli/dcoscli/package/main.py (hunk 9 above)
Comment (tamarrow) [inline cli/dcoscli/package/main.py]: done

File: cli/dcoscli/package/main.py (hunk 7 above)
Comment (tamarrow) [inline cli/dcoscli/package/main.py]: done

File: cli/dcoscli/package/main.py (hunk 6 above)
Comment (tamarrow) [inline cli/dcoscli/package/main.py]: done

File: cli/dcoscli/package/main.py (hunk 10 above)
Comment (tamarrow) [inline cli/dcoscli/package/main.py]: done

File: dcos/subcommand.py (hunk 1 above)
Comment (tamarrow) [inline dcos/subcommand.py]: done

File: dcos/subcommand.py (hunk 2 above)
Comment (tamarrow) [inline dcos/subcommand.py]: done

File: dcos/subcommand.py (hunk 3 above)
Comment (tamarrow) [inline dcos/subcommand.py]: done

File: dcos/subcommand.py (hunk 4 above)
Comment (tamarrow) [inline dcos/subcommand.py]: done

File: dcos/subcommand.py (hunk 5 above)
Comment (tamarrow) [inline dcos/subcommand.py]: The new function is just a part of `list_paths` that I separated out.

**Comment by tamarrow** [comment]
@mgummelt updated based on feedback

File: dcos/subcommand.py [hunk 11]
```diff
@@ -39,15 +37,38 @@ def command_executables(subcommand, dcos_path):
...
+        if (filename.startswith(constants.DCOS_COMMAND_PREFIX) and
+                _is_executable(path)):
+
+            executables.append(path)
+
+    return executables
+
+
+def list_paths():
+    """List the absolute path to executable dcos subcommand programs.
```
Comment (mgummelt) [inline dcos/subcommand.py]: I actually don't think "real" meant absolute.  I think it meant "no symbolic links": https://docs.python.org/2/library/os.path.html#os.path.realpath

Maybe change it back

File: cli/dcoscli/package/main.py [hunk 12]
```diff
@@ -348,6 +349,14 @@ def _install(package_name, options_path, app_id, cli, app, yes):
 
         subcommand.install(pkg, pkg_version, options)
 
+        subcommand_paths = subcommand.get_package_commands(package_name)
+        new_commands = [os.path.basename(p).replace('-', ' ', 1)
+                        for p in subcommand_paths]
+
+        if new_commands:
+            commands = ', '.join(new_commands)
+            emitter.publish("New command(s) available: {}".format(commands))
```
Comment (mgummelt) [inline cli/dcoscli/package/main.py]: You could use the length of the list to avoid "(s)"

**Comment by mgummelt** [comment]
minor comments then LGTM

File: dcos/subcommand.py (hunk 11 above)
Comment (tamarrow) [inline dcos/subcommand.py]: done

File: cli/dcoscli/package/main.py (hunk 12 above)
Comment (tamarrow) [inline cli/dcoscli/package/main.py]: done

**Comment by tamarrow** [comment]
fixed, thanks!

**Pull request merged**